### ISSUE TITLE
Warn when iterating an array of ranges

### ISSUE THREAD

**Issue by teor2345** [rust-lang/rust-clippy#7125]
### What it does

Warns users if they accidentally iterate range struct items, rather than iterating through the indexes in the range:

Suggests:
- iterating through the indexes in the range, or
- disambiguating the range struct iteration by providing a range type.

This lint will become more important once `rustc` allows iterating arrays by value, because that allows code like:
```rust
for _ in [0..5] {} // iterates through 1 range, not 5 indexes
```

For context, see https://github.com/rust-lang/rust/pull/84147#discussion_r612045984

### Categories (optional)

- Kind: `clippy::correctness`

*What is the advantage of the recommended code over the original code*

The recommended code either:
- does what the user actually intended, or
- makes it clear that the user intended something really unusual (iterating ranges themselves, not their indexes).

### Drawbacks

Expressing iterating through an array of ranges becomes more verbose. But this disambiguation makes the intent of the code much clearer.

### Example

```rust
for _ in [0..5] {} // iterates through 1 range, clippy warning
```

Is probably intended to be:

```rust
for _ in 0..5 {} // iterates through 5 indexes
```

And if range iteration is genuinely intended, it could be disambiguated as:

```rust
for _: Range<_> in [0..5] {} // iterates through 1 range, no clippy warning
```

**Comment by scottmcm** [comment]
nit: you can't actually annotate a type for the pattern in a `for` loop -- it's not supported in the grammar.

**Comment by matthiaskrgr** [comment]
i guess this should only lint on arrays of len 1?

what about code like this:
```
pub fn main() {
    for r in [0..7, 8..5, 37..42] {
        println!("{:?}",r);
    }
}
```

**Comment by llogiq** [comment]
Shouldn't that one be caught by `single_element_loop`?

**Comment by llogiq** [comment]
Looking into that, it catches `for r in &[0..5]`, but not (yet) `for r in [0..5].iter()`.

**Comment by llogiq** [comment]
I'll extend `single_element_loop` to also catch the `.iter()` variant.

**Comment by camsteffen** [comment]
What if we just lint `for _ in [any_single_element]` since that is never really useful.

**Comment by llogiq** [comment]
Yes, but that appears to only work if the element is an `ExprPath` for some reason.

**Comment by pitaj** [comment]
It appears that, actually, `single_element_loop` won't do anything with a straight array passed in. I've made some changes in #8616 that should cover that case. It also expands coverage to `.iter_mut()` and `.into_iter()`, and wraps in parens if necessary.

**Comment by pitaj** [comment]
We probably still want to cover the range case specially at some point, to at least give a better diagnostic.

**Comment by christophbeberweil** [comment]
Hi, @ThinkerDreamer and myself worked on a pull request  #11862 that lints the array itself, if it contains exactly one range. In this case, the lint suggests that the user wanted to iterate over the single range instead and suggests to rewrite the for loop accordingly.

**Comment by teor2345** [comment]
I think we can call this closed by #11862.
